ops/framework.py

```diff
@@ -32,6 +32,27 @@
 from ops import charm
 from ops.storage import NoSnapshotError, SQLiteStorage
 
+if typing.TYPE_CHECKING:
+    from ops.charm import CharmMeta
+    from ops.model import Model
+    from pathlib import Path
+
+    try:
+        from typing import Type, Protocol
+    except ModuleNotFoundError:
+        from typing_extensions import Type, Protocol
+
+    class _HandleLike(Protocol):
+        handle: "Handle"
```

Comment:
Unfortunately I believe the syntax handling for these "inline" variable type annotations was only [added in Python 3.6](https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-pep526), so we'll have to wait till we drop Python 3.5 for those.

Reply:
Yes indeed, this PR is likely to become a bit of a rabbit hole (I started once before...!).

We might consider using [pyright](https://github.com/microsoft/pyright). I understand it produces slightly friendlier error messages and runs a bit quicker than `mypy`, though honestly I'm ambivalent so long as it works for the Charm Tech folks.

Reply:
Uh right! forgot about that detail. Will fix it.
Before I dive into this further, however: are there any chance this will get merged, or is there no agreement that this is necessary/useful?

Reply:
I thought, naively, that since there are no changes to the logic, we could do it all at once. But let's see how far I get with this branch then we can choose to split the PR if we see it's too large